pkg/rpm/database.go

```diff
@@ -0,0 +1,285 @@
+		// Remove temporary directory if returning on errors.
+		if err != nil {
+			_ = os.RemoveAll(dbDir)
+		}
+	}()
+	for name, data := range dbFiles {
+		dbFilename := filepath.Join(dbDir, name)
+		err = os.WriteFile(dbFilename, data.Contents, 0700)
+		if err != nil {
+			logrus.WithError(err).Error("failed to create rpm database file")
+			return nil, commonerr.ErrFilesystem
+		}
+	}
+	return &rpmDatabase{
+		dbPath: dbDir,
+	}, nil
+}
+
+// QueryAll starts a query for all packages in the RPM database, returns a query
+// iterator. A query is actually performed by an underlying rpm sub-process, and
+// return is retrieved after parsing its output and status.
+func (d *rpmDatabase) QueryAll(opts QueryOpts) *rpmDatabaseQuery {
+	queryFormat := queryFmt
+	if opts.Testing {
+		queryFormat = queryFmtTest
+	}
+	cmd := exec.Command(
+		"rpm",
+		`--dbpath`, d.dbPath,
+		`--query`,
+		`--all`,
+		`--queryformat`, queryFormat,
+	)
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		return &rpmDatabaseQuery{err: err}
+	}
+	var errBuffer bytes.Buffer
+	cmd.Stderr = &errBuffer
+	if err := cmd.Start(); err != nil {
+		utils.IgnoreError(stdoutPipe.Close)
+		return &rpmDatabaseQuery{err: err}
+	}
+	return &rpmDatabaseQuery{
+		rpmWait:    cmd.Wait,
+		rpmStderr:  &errBuffer,
+		rpmStdout:  stdoutPipe,
+		rpmScanner: bufio.NewScanner(stdoutPipe),
+	}
+}
+
+// ProvidesFile return true if a package provides the specified path in the RPM
+// database. If the path is relative, we assume its relative to the root
+// directory.
+func (d *rpmDatabase) ProvidesFile(path string) bool {
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	cmd := exec.Command(
+		"rpm",
+		`--dbpath`, d.dbPath,
+		`--query`,
+		`--whatprovides`, path,
+	)
+	if err := cmd.Run(); err != nil {
+		// When rpm does not provide a file, the expected exit status is 1. On non-zero
+		// we always default to returning the file is not provided, but anything other
+		// than 1 is considered unexpected.
+		if exitError, ok := err.(*exec.ExitError); ok && exitError.ExitCode() != 1 {
+			logrus.WithError(err).Errorf(
+				"unexpected exit status when querying %s is provided by an RPM package", path)
+		}
+		return false
+	}
+	// The rpm exited properly, which implies the file IS provided by an RPM
+	// package.
+	return true
+}
+
+// Delete removes all the RPM database files.
+func (d *rpmDatabase) Delete() error {
+	return os.RemoveAll(d.dbPath)
+}
+
+// Next retrieves the next item in the query and make it available through the Package call.
+func (q *rpmDatabaseQuery) Next() bool {
+	if q.err != nil || q.rpmScanStopped {
+		return false
+	}
+	q.nextPackage = rpmPackage{}
+	for i := 0; q.rpmScanner.Scan(); i++ {
+		line := strings.TrimSpace(q.rpmScanner.Text())
+		if line == "" || strings.HasPrefix(line, "(none)") {
+			continue
+		}
+		if line == "." {
+			// Reached package delimiter. Ensure the current package is well-formed. If it
+			// is, add it to the return slice.
```

Comment:
forgot to mention this before, but let's update this comment since we no longer return a slice